### PR TITLE
Merge feature/viewmodelfactory-non-generic-methods into develop

### LINKED ISSUE #136: ViewModelFactory add non generic support

### PR DESCRIPTION
**Pull request by LegendaryB**
Closes #136

### CODE REVIEW

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 1]
```diff
@@ -142,7 +141,133 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
+            bool initialize = true)
+        {
+            var interfaceType = typeof(IViewModel);
+
+            try
+            {
+                if (!interfaceType.IsAssignableFrom(viewModelType))
+                {
+                    throw new InvalidOperationException(
+                        $"The type {viewModelType.Name} must implement interface type {interfaceType.Name}");
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:216]: Codacy found an issue: [Refactor this method to add validation of parameter 'viewModelType' before using it.](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

**Comment by Pantheas** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 6
           

Complexity increasing per file
==============================
- src/MVVM/Factories/ViewModelFactory.cs  1
         

Clones added
============
- src/MVVM/Factories/ViewModelFactory.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060&bid=18874753)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 2]
```diff
@@ -54,19 +54,18 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
-        /// Creates a new ViewModel instance and returns it
-        /// InitializeAsync is invoked in the background
+        /// Creates a new ViewModel instance and optionally invokes the <see cref="IViewModel.InitializeAsync"/> method.
         /// </summary>
-        /// <typeparam name="TViewModelInterface">Type of the ViewModel to create</typeparam>
-        /// <param name="initialize">Should InitializeAsync() get called after getting the instance? <see cref="IViewModel"/></param>
-        /// <returns>ViewModel instance of the given type</returns>
-        public static TViewModelInterface Resolve<TViewModelInterface>(
+        /// <typeparam name="TViewModel">Type of the ViewModel to create</typeparam>
+        /// <param name="initialize">Specifies if <see cref="IViewModel.InitializeAsync"/> should get called after instance creation.</param>
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:60]: Codacy found an issue: [Split this 141 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 3]
```diff
@@ -142,7 +141,133 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
+                    innerException);
+            }
+        }
+
+        /// <summary>
+        /// Creates a new ViewModel instance and invokes the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method.
+        /// </summary>
+        /// <typeparam name="TModel">Type of the parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</typeparam>
+        /// <param name="viewModelType"></param>
+        /// <param name="model">Parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</param>
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:245]: Codacy found an issue: [Split this 140 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 4]
```diff
@@ -122,18 +121,18 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
         /// <summary>
-        /// Creates a new ViewModel instance, invokes the InitializeAsync method using the parameter and returns the initialized instance
+        /// Creates a new ViewModel instance and invokes the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method.
         /// </summary>
-        /// <typeparam name="TViewModelInterface">Type of the ViewModel to create</typeparam>
-        /// <typeparam name="TModel">Type of the parameter that will be used for initialization</typeparam>
-        /// <returns>ViewModel instance of the given type</returns>
-        public static async Task<TViewModelInterface> ResolveAsync<TViewModelInterface, TModel>(
+        /// <typeparam name="TViewModel">Type of the ViewModel to create</typeparam>
+        /// <typeparam name="TModel">Type of the parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</typeparam>
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:127]: Codacy found an issue: [Split this 161 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 5]
```diff
@@ -88,18 +87,18 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
+        /// Creates a new ViewModel instance and optionally invokes the <see cref="IViewModel.InitializeAsync"/> method.
         /// </summary>
-        /// <typeparam name="TViewModelInterface">Type of the ViewModel to create</typeparam>
-        /// <param name="initialize">Should InitializeAsync() get called after getting the instance? <see cref="IViewModel"/></param>
-        /// <returns>ViewModel instance of the given type</returns>
-        public static async Task<TViewModelInterface> ResolveAsync<TViewModelInterface>(
+        /// <typeparam name="TViewModel">Type of the ViewModel to create</typeparam>
+        /// <param name="initialize">Specifies if <see cref="IViewModel.InitializeAsync"/> should get called after instance creation.</param>
+        /// <returns>A ViewModel instance of the given type</returns>
+        public static async Task<TViewModel> ResolveAsync<TViewModel>(
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:95]: Codacy found an issue: [All 'ResolveAsync' method overloads should be adjacent.](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 6]
```diff
@@ -54,19 +54,18 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
+        /// Creates a new ViewModel instance and optionally invokes the <see cref="IViewModel.InitializeAsync"/> method.
         /// </summary>
-        /// <typeparam name="TViewModelInterface">Type of the ViewModel to create</typeparam>
-        /// <param name="initialize">Should InitializeAsync() get called after getting the instance? <see cref="IViewModel"/></param>
-        /// <returns>ViewModel instance of the given type</returns>
-        public static TViewModelInterface Resolve<TViewModelInterface>(
+        /// <typeparam name="TViewModel">Type of the ViewModel to create</typeparam>
+        /// <param name="initialize">Specifies if <see cref="IViewModel.InitializeAsync"/> should get called after instance creation.</param>
+        /// <returns>A ViewModel instance of the given type</returns>
+        public static TViewModel Resolve<TViewModel>(
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:62]: Codacy found an issue: [All 'Resolve' method overloads should be adjacent.](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

**Comment by LegendaryB** [comment]
Leave open, identified a bug. Will fix asap

**Review by Pantheas — CHANGES_REQUESTED** [review]
Fix Codacy issues, I will approve afterwards :)

**Comment by Pantheas** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 4
           

Complexity increasing per file
==============================
- src/MVVM/Factories/ViewModelFactory.cs  1
         

Clones added
============
- src/MVVM/Factories/ViewModelFactory.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060&bid=18874753)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 7]
```diff
@@ -121,19 +204,62 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
-        /// Creates a new ViewModel instance, invokes the InitializeAsync method using the parameter and returns the initialized instance
+        /// Creates a new ViewModel instance and invokes the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method.
         /// </summary>
-        /// <typeparam name="TViewModelInterface">Type of the ViewModel to create</typeparam>
-        /// <typeparam name="TModel">Type of the parameter that will be used for initialization</typeparam>
-        /// <returns>ViewModel instance of the given type</returns>
-        public static async Task<TViewModelInterface> ResolveAsync<TViewModelInterface, TModel>(
+        /// <typeparam name="TModel">Type of the parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</typeparam>
+        /// <param name="viewModelType"></param>
+        /// <param name="model">Parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</param>
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:246]: Codacy found an issue: [Split this 140 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 8]
```diff
@@ -88,18 +87,102 @@ internal static INavigationService TryResolveNavigationServiceInstance()
...
+        {
+            var interfaceType = typeof(IViewModel);
+
             try
             {
-                var instance = iocContainer.Resolve<TViewModelInterface>();
+                if (!interfaceType.IsAssignableFrom(viewModelType))
+                {
+                    throw new InvalidOperationException(
+                        $"The type {viewModelType.Name} must implement interface type {interfaceType.Name}");
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:182]: Codacy found an issue: [Refactor this method to add validation of parameter 'viewModelType' before using it.](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs [hunk 9]
```diff
@@ -121,19 +204,62 @@ internal static INavigationService TryResolveNavigationServiceInstance()
         }
 
 
+
+        /// <summary>
+        /// Creates a new ViewModel instance and invokes the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method.
+        /// </summary>
+        /// <typeparam name="TViewModel">Type of the ViewModel to create</typeparam>
+        /// <typeparam name="TModel">Type of the parameter that will be passed in the <see cref="IViewModel{TModel}.InitializeAsync(TModel)"/> method</typeparam>
```
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:212]: Codacy found an issue: [Split this 161 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

File: src/MVVM/Factories/ViewModelFactory.cs (hunk 2 above)
Comment (Pantheas) [inline src/MVVM/Factories/ViewModelFactory.cs:60]: Codacy found an issue: [Split this 141 characters long line (which is greater than 130 authorized).](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060)

**Comment by Pantheas** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
           

Complexity increasing per file
==============================
- src/MVVM/Factories/ViewModelFactory.cs  1
         

Clones added
============
- src/MVVM/Factories/ViewModelFactory.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/UltimateCodeMonkeys/CodeMonkeysMVVM/pullRequest?prid=5767060&bid=18874753)

**Pull request merged**